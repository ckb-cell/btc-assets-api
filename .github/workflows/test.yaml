---
name: Unit Tests

on:
  pull_request:
    branches:
      - main
      - develop

jobs:
  test:
    runs-on: ubuntu-latest

    services:
      redis:
        image: redis
        options: >-
          --health-cmd "redis-cli ping"
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5
        ports:
          - 6379:6379

    steps:
      - name: Checkout Code
        uses: actions/checkout@v3

      - name: Install PNPM
        uses: pnpm/action-setup@v3
        with:
          version: 8

      - name: Use Node.js
        uses: actions/setup-node@v3
        with:
          node-version: "21"
          cache: "pnpm"

      - name: Install dependencies
        run: pnpm install

      - name: Run Unit Tests
        env:
          JWT_SECRET: ${{ secrets.JWT_SECRET }}
          BITCOIN_JSON_RPC_URL: ${{ secrets.BITCOIN_JSON_RPC_URL }}
          BITCOIN_JSON_RPC_USERNAME: ${{ secrets.BITCOIN_JSON_RPC_USERNAME }}
          BITCOIN_JSON_RPC_PASSWORD: ${{ secrets.BITCOIN_JSON_RPC_PASSWORD }}
          BITCOIN_ELECTRS_API_URL: ${{ secrets.BITCOIN_ELECTRS_API_URL }}
          BITCOIN_SPV_SERVICE_URL: ${{ secrets.BITCOIN_SPV_SERVICE_URL }}
          CKB_RPC_URL: ${{ secrets.CKB_RPC_URL }}
          CKB_INDEXER_URL: ${{ secrets.CKB_INDEXER_URL }}
          PAYMASTER_PRIVATE_KEY: ${{ secrets.PAYMASTER_PRIVATE_KEY }}
          REDIS_URL: redis://localhost:6379
          CI_REDIS_URL: redis://localhost:6379
        run: pnpm test
