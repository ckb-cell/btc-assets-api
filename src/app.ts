import { FastifyInstance } from 'fastify';
import { AxiosError } from 'axios';
import bitcoinRoutes from './routes/bitcoin';
import { env } from './env';
import tokenRoutes from './routes/token';
import swaggerPlugin from './plugins/swagger';
import jwtPlugin from './plugins/jwt';
import cachePlugin from './plugins/cache';
import rateLimitPlugin from './plugins/rate-limit';
import * as Sentry from '@sentry/node';

if (env.SENTRY_DSN_URL && env.NODE_ENV !== 'development') {
  Sentry.init({
    dsn: env.SENTRY_DSN_URL,
    tracesSampleRate: 1.0,
  });
}

async function routes(fastify: FastifyInstance) {
  fastify.register(rateLimitPlugin);
  fastify.register(swaggerPlugin);
  fastify.register(jwtPlugin);
  fastify.register(cachePlugin);

  fastify.register(tokenRoutes, { prefix: '/token' });
  fastify.register(bitcoinRoutes, { prefix: '/bitcoin/v1' });

  fastify.setErrorHandler((error, _, reply) => {
    Sentry.captureException(error);
    if (error instanceof AxiosError) {
      reply
        .status(error.response?.status || 500)
        .send({ ok: false, error: error.response?.data.error ?? error.message });
      return;
    }
    reply
      .status(error.statusCode ?? 500)
      .send({ ok: false, statusCode: error.statusCode, message: error.message });
  });
}

export default routes;
