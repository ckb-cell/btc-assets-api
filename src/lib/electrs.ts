import axios, { AxiosInstance } from 'axios';

export interface UTXO {
  txid: string;
  vout: number;
  value: number;
  status: {
    confirmed: boolean;
    block_height: number;
    block_hash: string;
    block_time: number;
  };
}

export interface Balance {
  address: string;
  satoshi: number;
  pendingSatoshi: number;
  utxoCount: number;
}

export default class ElectrsAPI {
  private request: AxiosInstance;

  constructor(baseURL: string) {
    this.request = axios.create({
      baseURL,
    });
  }

  async getUtxoByAddress(address: string): Promise<UTXO[]> {
    const response = await this.request.get(`/address/${address}/utxo`);
    return response.data;
  }

  async getBalanceByAddress(address: string): Promise<Balance> {
    const utxos = await this.getUtxoByAddress(address);
    return utxos.reduce(
      (acc, utxo) => {
        if (utxo.status.confirmed) {
          acc.satoshi += utxo.value;
        } else {
          acc.pendingSatoshi += utxo.value;
        }
        return acc;
      },
      {
        address,
        satoshi: 0,
        pendingSatoshi: 0,
        utxoCount: utxos.length,
      },
    );
  }
}
