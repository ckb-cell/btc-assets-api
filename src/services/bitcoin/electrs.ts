import axios, { AxiosInstance } from 'axios';
import { Cradle } from '../../container';
import { IBitcoinDataProvider } from './interface';
import { Block, RecommendedFees, Transaction, UTXO } from './schema';

export class ElectrsClient implements IBitcoinDataProvider {
  private request: AxiosInstance;

  constructor(cradle: Cradle) {
    this.request = axios.create({
      baseURL: cradle.env.BITCOIN_ELECTRS_API_URL,
    });
  }

  public async getFeesRecommended(): Promise<RecommendedFees> {
    throw new Error('ElectrsClient does not support getFeesRecommended');
  }

  public async postTx({ txhex }: { txhex: string }) {
    const response = await this.request.post('/tx', txhex);
    return response.data;
  }

  public async getAddressTxsUtxo({ address }: { address: string }) {
    const response = await this.request.get<UTXO[]>(`/address/${address}/utxo`);
    return response.data;
  }

  public async getAddressTxs({ address, after_txid }: { address: string; after_txid?: string }) {
    let url = `/address/${address}/txs`;
    if (after_txid) {
      url += `?after_txid=${after_txid}`;
    }
    const response = await this.request.get<Transaction[]>(url);
    return response.data.map((tx) => Transaction.parse(tx));
  }

  public async getTx({ txid }: { txid: string }) {
    const response = await this.request.get<Transaction>(`/tx/${txid}`);
    return Transaction.parse(response.data);
  }

  public async getTxHex({ txid }: { txid: string }) {
    const response = await this.request.get<string>(`/tx/${txid}/hex`);
    return response.data;
  }

  public async getBlock({ hash }: { hash: string }) {
    const response = await this.request.get<Block>(`/block/${hash}`);
    return Block.parse(response.data);
  }

  public async getBlockHeight({ height }: { height: number }) {
    const response = await this.request.get<string>(`/block-height/${height}`);
    return response.data;
  }

  public async getBlockHeader({ hash }: { hash: string }) {
    const response = await this.request.get<string>(`/block/${hash}/header`);
    return response.data;
  }

  public async getBlockTxids({ hash }: { hash: string }) {
    const response = await this.request.get<string[]>(`/block/${hash}/txids`);
    return response.data;
  }

  public async getBlocksTipHash() {
    const response = await this.request.get<string>('/blocks/tip/hash');
    return response.data;
  }
}
