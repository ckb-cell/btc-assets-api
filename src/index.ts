import process from 'node:process';
import { env } from './env';
import { buildFastify } from './app';

const port = parseInt(env.PORT || '3000', 10);
const host = env.ADDRESS || '0.0.0.0';

const app = buildFastify();

app.listen({ port, host }, (err, address) => {
  if (err) {
    console.error(err);
    process.exit(1);
  }

  // eslint-disable-next-line no-console
  console.log(`Server listening at ${address}`);
});
