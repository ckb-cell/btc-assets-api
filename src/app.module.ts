import { Module } from '@nestjs/common';
import { AppService } from './app.service';
import { TokenModule } from './modules/token/token.module';
import { BitcoinModule } from './modules/bitcoin/bitcoin.module';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { envSchema } from './env';
import { SentryModule } from '@ntegral/nestjs-sentry';
import { nodeProfilingIntegration } from '@sentry/profiling-node';
import { CacheModule } from '@nestjs/cache-manager';
import type { RedisClientOptions } from 'redis';
import * as redisStore from 'cache-manager-redis-store';
import { StoreConfig } from 'cache-manager';

@Module({
  imports: [
    ConfigModule.forRoot({
      isGlobal: true,
      validate: envSchema.parse,
    }),
    CacheModule.registerAsync<RedisClientOptions>({
      isGlobal: true,
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: async (configService: ConfigService) =>
        ({
          store: redisStore,
          url: configService.get('REDIS_URL'),
          ttl: 10,
        }) as StoreConfig,
    }),
    SentryModule.forRootAsync({
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: async (configService: ConfigService) => ({
        dsn: configService.get('SENTRY_DSN'),
        environment: configService.get('NODE_ENV'),
        tracesSampleRate: 0.5,
        profilesSampleRate: 0.5,
        integrations: [nodeProfilingIntegration()],
        logLevels:
          configService.get('NODE_ENV') === 'production'
            ? ['warn', 'error']
            : ['debug', 'log', 'warn', 'error'],
        beforeSend: (event) => {
          if (event.level === 'error') {
            return event;
          }
          return null;
        },
      }),
    }),
    TokenModule,
    BitcoinModule,
  ],
  controllers: [],
  providers: [AppService],
})
export class AppModule {}
