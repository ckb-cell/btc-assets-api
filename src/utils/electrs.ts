import axios, { AxiosInstance, AxiosResponse } from 'axios';
import { Transaction, Block } from 'bitcoinjs-lib';
import { Cradle } from '../container';
import { UTXO } from '../routes/bitcoin/types';
import { addLoggerInterceptor } from './interceptors';

export default class ElectrsClient {
  private request: AxiosInstance;

  constructor({ env, logger }: Cradle) {
    this.request = axios.create({
      baseURL: env.BITCOIN_ELECTRS_API_URL,
    });
    addLoggerInterceptor(this.request, logger);
  }

  private async get<T>(path: string): Promise<AxiosResponse<T>> {
    const response = await this.request.get(path);
    return response;
  }

  public async sendRawTransaction(hex: string) {
    const response = await this.request.post('/tx', hex);
    return response.data;
  }

  public async getUtxoByAddress(address: string) {
    const response = await this.get<UTXO[]>(`/address/${address}/utxo`);
    return response.data;
  }

  public async getTransactionsByAddress(address: string) {
    const response = await this.get<Transaction[]>(`/address/${address}/txs`);
    return response.data;
  }

  public async getTransaction(txid: string) {
    const response = await this.get<Transaction>(`/tx/${txid}`);
    return response.data;
  }

  public async getTransactionHex(txid: string) {
    const response = await this.get<string>(`/tx/${txid}/hex`);
    return response.data;
  }

  public async getBlockByHash(hash: string) {
    const response = await this.get<Block>(`/block/${hash}`);
    return response.data;
  }

  public async getBlockHashByHeight(height: number) {
    const response = await this.get<string>(`/block-height/${height}`);
    return response.data;
  }

  public async getBlockHeaderByHash(hash: string) {
    const response = await this.get<string>(`/block/${hash}/header`);
    return response.data;
  }

  public async getBlockTxIdsByHash(hash: string) {
    const response = await this.get<string[]>(`/block/${hash}/txids`);
    return response.data;
  }

  public async getBlocksTipHash() {
    const response = await this.get<string>('/blocks/tip/hash');
    return response.data;
  }
}
