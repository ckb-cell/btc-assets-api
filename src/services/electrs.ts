import axios, { AxiosInstance, AxiosResponse } from 'axios';
import { BlockType, TransactionType, UTXOType } from '../routes/bitcoin/types';
import * as Sentry from '@sentry/node';
import { Cradle } from '../container';
import { addLoggerInterceptor } from '../utils/interceptors';

export default class ElectrsAPI {
  private request: AxiosInstance;

  constructor({ env, logger }: Cradle) {
    this.request = axios.create({
      baseURL: env.BITCOIN_ELECTRS_API_URL,
    });
    addLoggerInterceptor(this.request, logger);
  }

  private async get<T>(path: string): Promise<AxiosResponse<T>> {
    return Sentry.startSpan({ op: this.constructor.name, name: path }, async () => {
      const response = await this.request.get(path);
      return response;
    });
  }

  // https://github.com/blockstream/esplora/blob/master/API.md#get-addressaddressutxo
  public async getUtxoByAddress(address: string) {
    const response = await this.get<UTXOType[]>(`/address/${address}/utxo`);
    return response.data;
  }

  // https://github.com/blockstream/esplora/blob/master/API.md#get-addressaddresstxs
  public async getTransactionsByAddress(address: string) {
    const response = await this.get<TransactionType[]>(`/address/${address}/txs`);
    return response.data;
  }

  // https://github.com/blockstream/esplora/blob/master/API.md#get-txtxid
  public async getTransaction(txid: string) {
    const response = await this.get<TransactionType>(`/tx/${txid}`);
    return response.data;
  }

  // https://github.com/blockstream/esplora/blob/master/API.md#get-blockhash
  public async getBlockByHash(hash: string) {
    const response = await this.get<BlockType>(`/block/${hash}`);
    return response.data;
  }

  // https://github.com/blockstream/esplora/blob/master/API.md#get-block-heightheight
  public async getBlockByHeight(height: number) {
    const response = await this.get<string>(`/block-height/${height}`);
    return response.data;
  }

  // https://github.com/blockstream/esplora/blob/master/API.md#get-blockhashheader
  public async getBlockHeaderByHash(hash: string) {
    const response = await this.get<string>(`/block/${hash}/header`);
    return response.data;
  }
}
