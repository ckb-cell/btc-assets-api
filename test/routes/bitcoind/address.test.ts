import { describe, expect, test, beforeEach, vi } from 'vitest';
import { buildFastify } from '../../../src/app';
import { ElectrsAPIError, ElectrsAPIErrorCode, ElectrsErrorMessage } from '../../../src/services/electrs';
import { afterEach } from 'node:test';

let token: string;

describe('/bitcoin/v1/address', () => {
  beforeEach(async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const response = await fastify.inject({
      method: 'POST',
      url: '/token/generate',
      payload: {
        app: 'test',
        domain: 'test.com',
      },
    });
    const data = response.json();
    token = data.token;

    await fastify.close();
  });

  afterEach(() => {
    vi.resetAllMocks();
  });

  test('Get address balance', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const response = await fastify.inject({
      method: 'GET',
      url: '/bitcoin/v1/address/tb1qm4eyx777203zmajlawz958wn27z08envm2jelm/balance',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = response.json();

    expect(response.statusCode).toBe(200);
    expect(data).toMatchSnapshot();

    await fastify.close();
  });

  test('Get address balance with min_satoshi param', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const response = await fastify.inject({
      method: 'GET',
      url: '/bitcoin/v1/address/tb1qm4eyx777203zmajlawz958wn27z08envm2jelm/balance?min_satoshi=10000',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = response.json();

    expect(response.statusCode).toBe(200);
    expect(data).toMatchSnapshot();

    await fastify.close();
  });

  test('Get address balance with invalid address', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const response = await fastify.inject({
      method: 'GET',
      url: '/bitcoin/v1/address/tb1qlrg2mhyxrq7ns5rpa6qvrvttr9674n6z0try/balance',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = response.json();

    expect(response.statusCode).toBe(400);
    expect(data.message).toBe('Invalid bitcoin address');

    await fastify.close();
  });

  test('Get address unspent transaction outputs', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const electrs = fastify.container.resolve('electrs');
    const originalGetUtxoByAddress = electrs.getUtxoByAddress;
    vi.spyOn(electrs, 'getUtxoByAddress').mockResolvedValue([
      {
        txid: '9706131c1e327a068a6aafc16dc69a46c50bc7c65f180513896bdad39a6babfc',
        vout: 0,
        status: {
          confirmed: true,
        },
        value: 100000,
      },
      {
        txid: '1706131c1e327a068a6aafc16dc69a46c50bc7c65f180513896bdad39a6babfc',
        vout: 0,
        status: {
          confirmed: false,
        },
        value: 100000,
      },
    ]);

    const response = await fastify.inject({
      method: 'GET',
      url: '/bitcoin/v1/address/tb1qm4eyx777203zmajlawz958wn27z08envm2jelm/unspent',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = response.json();
    electrs.getUtxoByAddress = originalGetUtxoByAddress;

    expect(response.statusCode).toBe(200);
    expect(data.length).toBe(1);

    await fastify.close();
  });

  test('Get address unspent transaction outputs with unconfirmed', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const electrs = fastify.container.resolve('electrs');
    const originalGetUtxoByAddress = electrs.getUtxoByAddress;
    vi.spyOn(electrs, 'getUtxoByAddress').mockResolvedValue([
      {
        txid: '9706131c1e327a068a6aafc16dc69a46c50bc7c65f180513896bdad39a6babfc',
        vout: 0,
        status: {
          confirmed: true,
        },
        value: 100000,
      },
      {
        txid: '1706131c1e327a068a6aafc16dc69a46c50bc7c65f180513896bdad39a6babfc',
        vout: 0,
        status: {
          confirmed: false,
        },
        value: 100000,
      },
    ]);

    const response = await fastify.inject({
      method: 'GET',
      url: '/bitcoin/v1/address/tb1qm4eyx777203zmajlawz958wn27z08envm2jelm/unspent?only_confirmed=0',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = response.json();
    electrs.getUtxoByAddress = originalGetUtxoByAddress;

    expect(response.statusCode).toBe(200);
    expect(data.length).toBe(2);

    await fastify.close();
  });

  test('Get address unspent transaction outputs throw too many', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const response = await fastify.inject({
      method: 'GET',
      url: '/bitcoin/v1/address/tb1qcq670zweall6zz4f96flfrefhr8myfxz9ll9l2/unspent',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = response.json();

    expect(response.statusCode).toBe(500);
    expect(data.code).toEqual(ElectrsAPIErrorCode.TooManyUtxos);
    expect(data.message).toEqual('Too many unspent transaction outputs');

    await fastify.close();
  });

  test('Get address transactions', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const response = await fastify.inject({
      method: 'GET',
      url: '/bitcoin/v1/address/tb1qm4eyx777203zmajlawz958wn27z08envm2jelm/txs',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = response.json();

    expect(response.statusCode).toBe(200);
    expect(data).toMatchSnapshot();

    await fastify.close();
  });
});
