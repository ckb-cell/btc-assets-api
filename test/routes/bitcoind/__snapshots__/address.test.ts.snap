// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`/bitcoin/v1/address > Get address transactions 1`] = `
[
  {
    "fee": 141,
    "locktime": 0,
    "size": 223,
    "status": {
      "block_hash": "000000000000000cef4a1d6264fe63f543128518a466d31c7e2a8d6395b52522",
      "block_height": 2579043,
      "block_time": 1708580004,
      "confirmed": true,
    },
    "txid": "9706131c1e327a068a6aafc16dc69a46c50bc7c65f180513896bdad39a6babfc",
    "version": 2,
    "vin": [
      {
        "is_coinbase": false,
        "prevout": {
          "scriptpubkey": "0014f8d0addc86183d385061ee80c1b16b1975eacf42",
          "scriptpubkey_address": "tb1qlrg2mhyxrq7ns5rpa6qvrvttr9674n6z0trymp",
          "scriptpubkey_asm": "OP_0 OP_PUSHBYTES_20 f8d0addc86183d385061ee80c1b16b1975eacf42",
          "scriptpubkey_type": "v0_p2wpkh",
          "value": 181793,
        },
        "scriptsig": "",
        "scriptsig_asm": "",
        "sequence": 4294967295,
        "txid": "d0189f19978fc47ddfe33319d01a6a66dea5a521e1f37b7e73469a3549e81531",
        "vout": 0,
        "witness": [
          "3045022100f50e2e6ee5ea04e4f7029d29a28e80266969f91357055e1d53b0715b768c448102204ccd9dbb38d0cc007e07b9c8b4088d4fac4ab65fef49774bafe27acbfc3d111101",
          "02d05848540f152d730e272bc9628dd4e4a5f4126fdf118ad6a4fd6afd26b08313",
        ],
      },
    ],
    "vout": [
      {
        "scriptpubkey": "0014dd72437bde53e22df65feb845a1dd35784f3e66c",
        "scriptpubkey_address": "tb1qm4eyx777203zmajlawz958wn27z08envm2jelm",
        "scriptpubkey_asm": "OP_0 OP_PUSHBYTES_20 dd72437bde53e22df65feb845a1dd35784f3e66c",
        "scriptpubkey_type": "v0_p2wpkh",
        "value": 1000,
      },
      {
        "scriptpubkey": "0014f8d0addc86183d385061ee80c1b16b1975eacf42",
        "scriptpubkey_address": "tb1qlrg2mhyxrq7ns5rpa6qvrvttr9674n6z0trymp",
        "scriptpubkey_asm": "OP_0 OP_PUSHBYTES_20 f8d0addc86183d385061ee80c1b16b1975eacf42",
        "scriptpubkey_type": "v0_p2wpkh",
        "value": 180652,
      },
    ],
    "weight": 562,
  },
  {
    "fee": 291,
    "locktime": 2579039,
    "size": 341,
    "status": {
      "block_hash": "000000000000000cef4a1d6264fe63f543128518a466d31c7e2a8d6395b52522",
      "block_height": 2579043,
      "block_time": 1708580004,
      "confirmed": true,
    },
    "txid": "85fdce5f5d7fd3ff73ce70e3e0a786f50cc1124830cc07341738d76fa7c3a6a9",
    "version": 2,
    "vin": [
      {
        "is_coinbase": false,
        "prevout": {
          "scriptpubkey": "51200ee08c77cfb2d35e6cab21b88e97c3acc5c3a3888237f2a0a6cc8006d8283487",
          "scriptpubkey_address": "tb1ppmsgca70ktf4um9tyxuga97r4nzu8gugsgml9g9xejqqdkpgxjrskjx4wu",
          "scriptpubkey_asm": "OP_PUSHNUM_1 OP_PUSHBYTES_32 0ee08c77cfb2d35e6cab21b88e97c3acc5c3a3888237f2a0a6cc8006d8283487",
          "scriptpubkey_type": "v1_p2tr",
          "value": 978803861,
        },
        "scriptsig": "",
        "scriptsig_asm": "",
        "sequence": 4294967293,
        "txid": "e8b3e53f97f5c32a53b3af6c7b6dd3ad29439e41f09ad5bf14d6a970fc626826",
        "vout": 1,
        "witness": [
          "92aeef07a488114bf9d1383a67846e0d50d48023755bc835278fca2ed42ade5c7f2749de20909b43a02b8222fa1849505df98b7705d575e31db90fdf70400b68",
        ],
      },
    ],
    "vout": [
      {
        "scriptpubkey": "51209d18a7c0933c238e1af711891d288db56b64b4e263b33ffdf8e15d8320ff2bf6",
        "scriptpubkey_address": "tb1pn5v20syn8s3cuxhhzxy362ydk44kfd8zvwenll0cu9wcxg8l90mq80y9kr",
        "scriptpubkey_asm": "OP_PUSHNUM_1 OP_PUSHBYTES_32 9d18a7c0933c238e1af711891d288db56b64b4e263b33ffdf8e15d8320ff2bf6",
        "scriptpubkey_type": "v1_p2tr",
        "value": 978796570,
      },
      {
        "scriptpubkey": "5120c1976714c66503901967f48227db5c1e98945ddefca53d6c875c3bb50ca842dc",
        "scriptpubkey_address": "tb1pcxtkw9xxv5peqxt87jpz0k6ur6vfghw7ljjn6my8tsam2r9ggtwq32vtnk",
        "scriptpubkey_asm": "OP_PUSHNUM_1 OP_PUSHBYTES_32 c1976714c66503901967f48227db5c1e98945ddefca53d6c875c3bb50ca842dc",
        "scriptpubkey_type": "v1_p2tr",
        "value": 1000,
      },
      {
        "scriptpubkey": "001405515fbd2be2e4375262d609c3ddd4db63611e6f",
        "scriptpubkey_address": "tb1qq4g4l0ftutjrw5nz6cyu8hw5md3kz8n0hydu7e",
        "scriptpubkey_asm": "OP_0 OP_PUSHBYTES_20 05515fbd2be2e4375262d609c3ddd4db63611e6f",
        "scriptpubkey_type": "v0_p2wpkh",
        "value": 2000,
      },
      {
        "scriptpubkey": "5120e79527cd47ed7cfe1099b630a90da5aabc6c858b4c01f332e42b461bcb5ae043",
        "scriptpubkey_address": "tb1pu72j0n28a470uyyekcc2jrd9427xepvtfsqlxvhy9drphj66upps2c9a63",
        "scriptpubkey_asm": "OP_PUSHNUM_1 OP_PUSHBYTES_32 e79527cd47ed7cfe1099b630a90da5aabc6c858b4c01f332e42b461bcb5ae043",
        "scriptpubkey_type": "v1_p2tr",
        "value": 2000,
      },
      {
        "scriptpubkey": "0014f8d0addc86183d385061ee80c1b16b1975eacf42",
        "scriptpubkey_address": "tb1qlrg2mhyxrq7ns5rpa6qvrvttr9674n6z0trymp",
        "scriptpubkey_asm": "OP_0 OP_PUSHBYTES_20 f8d0addc86183d385061ee80c1b16b1975eacf42",
        "scriptpubkey_type": "v0_p2wpkh",
        "value": 1000,
      },
      {
        "scriptpubkey": "0014363e9d485ada999864c65c01f46dd2b7be6c35aa",
        "scriptpubkey_address": "tb1qxclf6jz6m2vesexxtsqlgmwjk7lxcdd2lt3wxs",
        "scriptpubkey_asm": "OP_0 OP_PUSHBYTES_20 363e9d485ada999864c65c01f46dd2b7be6c35aa",
        "scriptpubkey_type": "v0_p2wpkh",
        "value": 1000,
      },
    ],
    "weight": 1160,
  },
  {
    "fee": 14100,
    "locktime": 2579039,
    "size": 222,
    "status": {
      "block_hash": "000000000000001d5af479c1bcdc1202913a82c652ddc45b8bf7110e4977e4be",
      "block_height": 2579040,
      "block_time": 1708578362,
      "confirmed": true,
    },
    "txid": "d0189f19978fc47ddfe33319d01a6a66dea5a521e1f37b7e73469a3549e81531",
    "version": 2,
    "vin": [
      {
        "is_coinbase": false,
        "prevout": {
          "scriptpubkey": "00143e5700df69430bcf88d7c83980d09d4d5f88ec71",
          "scriptpubkey_address": "tb1q8etsphmfgv9ulzxhequcp5yaf40c3mr3vxdxhu",
          "scriptpubkey_asm": "OP_0 OP_PUSHBYTES_20 3e5700df69430bcf88d7c83980d09d4d5f88ec71",
          "scriptpubkey_type": "v0_p2wpkh",
          "value": 1818812120,
        },
        "scriptsig": "",
        "scriptsig_asm": "",
        "sequence": 4294967293,
        "txid": "e235e27e49adf6deee9e9e91450d3101ba5ff82b90a5584c056c0fdf0f3f9ec0",
        "vout": 0,
        "witness": [
          "304402204575aad8b34d55c4c4f814c46e741aab8b6c614dcfacf55172b174629e19558002205eca835e9696f0f17655e4458bf3012ac7021aa1b228a8c4fec5e1f2a8444edf01",
          "03a9c5a1c05b98a249cc54f9f96f0dbdaa7bd3110ab9f52bd8ee5ac9e0aad9c008",
        ],
      },
    ],
    "vout": [
      {
        "scriptpubkey": "0014f8d0addc86183d385061ee80c1b16b1975eacf42",
        "scriptpubkey_address": "tb1qlrg2mhyxrq7ns5rpa6qvrvttr9674n6z0trymp",
        "scriptpubkey_asm": "OP_0 OP_PUSHBYTES_20 f8d0addc86183d385061ee80c1b16b1975eacf42",
        "scriptpubkey_type": "v0_p2wpkh",
        "value": 181793,
      },
      {
        "scriptpubkey": "001424e9ecf624782719673221d69ea24b3aa76d22f3",
        "scriptpubkey_address": "tb1qyn57ea3y0qn3jeejy8tfagjt82nk6ghnr9583w",
        "scriptpubkey_asm": "OP_0 OP_PUSHBYTES_20 24e9ecf624782719673221d69ea24b3aa76d22f3",
        "scriptpubkey_type": "v0_p2wpkh",
        "value": 1818616227,
      },
    ],
    "weight": 561,
  },
]
`;
